pkg/ebpf/bpf_errors.go

```diff
@@ -0,0 +1,228 @@
+	}
+}
+
+func (b BPFError) Return() int64 {
+	return int64(b.ret)
+}
+
+func (b BPFError) ID() uint32 {
+	return uint32(b.id)
+}
+
+func (b BPFError) Type() BPFErrorType {
+	return BPFErrorType(b.id)
+}
+
+func (b BPFError) Line() uint32 {
+	return b.line
+}
+
+func (b BPFError) Function() []byte {
+	return b.function[:]
+}
+
+func (b BPFError) FunctionAsString() string {
+	errFunc := string(b.function[:])
+	errFuncEnd := strings.Index(errFunc, "\x00")
+	return errFunc[:errFuncEnd]
+}
+
+func (b BPFError) returnSize() int {
+	return int(unsafe.Sizeof(b.ret))
+}
+
+func (b BPFError) idSize() int {
+	return int(unsafe.Sizeof(b.id))
+}
+
+func (b BPFError) lineSize() int {
+	return int(unsafe.Sizeof(b.line))
+}
+
+func (b BPFError) functionSize() int {
+	return int(len(b.function))
+}
+
+func (b BPFError) Size() int {
+	return b.returnSize() + b.idSize() + b.lineSize() + b.functionSize()
+}
+
+// decode decodes rawData updating fields
+// - Fields must be passed as pointers
+// - The fields order matters
+func decode(rawData []byte, fields ...interface{}) error {
+	if len(rawData) == 0 {
+		return fmt.Errorf("decode: raw data is empty")
+	}
+	if len(fields) == 0 {
+		return fmt.Errorf("decode: fields must be set")
+	}
+
+	cursor := 0
+	delimiter := 0
+	for _, field := range fields {
+		updateCursor := func(v int) {
+			cursor += v
+		}
+		setDelimiter := func(v int) error {
+			if v > len(rawData) {
+				return fmt.Errorf("decode: delimiter %d out of bounds %d", v, len(rawData)-1)
+			}
+			delimiter = v
+
+			return nil
+		}
+
+		if cursor >= len(rawData) {
+			return fmt.Errorf("decode: cursor %d out of bounds %d - field %T not decoded", cursor, len(rawData)-1, field)
+		}
+
+		switch dest := field.(type) {
+		case *uint32:
+			typeSize := int(unsafe.Sizeof(*dest))
+			if err := setDelimiter(cursor + typeSize); err != nil {
+				return err
+			}
+			*dest = binary.LittleEndian.Uint32(rawData[cursor:delimiter])
+			updateCursor(typeSize)
+
+		case *uint64:
+			typeSize := int(unsafe.Sizeof(*dest))
+			if err := setDelimiter(cursor + typeSize); err != nil {
+				return err
+			}
+			*dest = binary.LittleEndian.Uint64(rawData[cursor:delimiter])
+			updateCursor(typeSize)
+
+		case []byte:
+			typeSize := len(dest)
+			if err := setDelimiter(cursor + typeSize); err != nil {
+				return err
+			}
+			copy(dest, rawData[cursor:delimiter])
+			updateCursor(typeSize)
+
+		default:
+			return fmt.Errorf("decode: type %[1]T not defined (%[1]v)", dest)
+		}
+	}
+
+	return nil
+}
+
+func (b *BPFError) Decode(rawData []byte) error {
+	if len(rawData) < b.Size() {
+		return fmt.Errorf("can't decode error raw data - buffer of %d should be of %d bytes", len(rawData), b.Size())
+	}
+
+	return decode(rawData,
+		&b.ret,
+		&b.id,
+		&b.line,
+		b.function[:],
+	)
+}
+
+func (t *Tracee) processBPFErrors() {
+	for {
+		select {
+		case rawData := <-t.bpfErrorsChannel:
+			if len(rawData) == 0 {
+				continue
+			}
+
+			bpfErr := NewBPFError()
```

Comment:
Since we export the struct anyway and only use this in decoding, I think the constructor is a bit redundant.

Reply:
I agree. At start this would ensure that the error id is set to BPF_ERR_UNSPEC, which is math.MaxUint32. Perhaps changing BPF_ERR_UNSPEC to 0 is the best option here.

Reply:
Yep, usually when you do these kind of golang "enums" you set the unspecified value to iota exactly to have it as the default value.